File: Multi_Cell_Julia/Cell_funcs.jl
using Random
using Distributions 
include("EE_Ovlap_Area.jl")

# Function to check if new cell collides with any other cell present
function noCollision(cells, cparams, cell_index, a, b, x, 
                    y, theta)

    # No need to check for initial cell                    
    if cell_index == 1
        return 1
    end               
    
    x1 = x
    y1 = y 
    a1 = a
    b1 = b 
    t1 = theta 

    for i in 1:cell_index
        x2 = cells[3*i+1]
        y2 = cells[3*i+2]
        a2 = cparams[4*i+1]
        b2 = cparams[4*i+2]
        t2 = cparams[4*i+3]

        #intersection area 
        intersec = ellipse_ellipse_overlap(a1, b1, x1, y1, t1, 
        a2, b2, x2, y2, t2)

        if intersec > 0.0
            return 0
        end
    end

    # cell doesn't overlaps with existing cells
    return 1
end                    

# A two cell preset
function two_cell_preset(a, b, Num)

    # x, y and dtheta D.O.F 
    cells = zeros(Float64, 3*Num)
    # a b theta and phase
    cparams = zeros(Float64, 4*Num)
    # Overlap array
    Ovlaps = zeros(Int64, Num, Num)

    # Cell 1
    ind = 1
    x = 20.
    y = 20.
    theta = 0.0
    cells[3*(ind-1)+1:(3*(ind-1)+3)] = Float64[x, y, 0.0]
    cparams[4*(ind-1)+1:(4*(ind-1)+4)] = Float64[a, b, theta, 1]

    # Cell 2
    ind = 2
    x = 18.
    y = 13.
    theta = pi/4
    cells[3*(ind-1)+1:(3*(ind-1)+3)] = Float64[x, y, 0.0]
    cparams[4*(ind-1)+1:(4*(ind-1)+4)] = Float64[a, b, theta, 1]

    return cells, cparams, Ovlaps
end

# Linear preset 
function linear_preset(a, b, Num)

    # x, y and dtheta D.O.F 
    cells = zeros(Float64, 3*Num)
    # a b theta and phase
    cparams = zeros(Float64, 4*Num)
    # Overlap array
    Ovlaps = zeros(Int64, Num, Num)
    
    # Linear in X - direction
    for ind in 1:Num
        cells[3*(ind-1)+1:(3*(ind-1)+3)] = Float64[20+2*(a-1)*ind, 10, 0.0]
        cparams[4*(ind-1)+1:(4*(ind-1)+4)] = Float64[a, b, 0.0, 1] 
    end    

    return cells, cparams, Ovlaps
end    

# Random configuration of non-overlapping cells 
function random_cells(L, a, b, Num)

    # x, y and dtheta D.O.F 
    cells = zeros(Float64, 3*Num)
    # a b theta and phase
    cparams = zeros(Float64, 4*Num)
    # Overlap array
    Ovlaps = zeros(Int64, Num, Num)

    # cell index 
    cell_ind = 0

    while (cell_ind <= Num)
        x = rand(Uniform(0.0, L))
        y = rand(Uniform(0.0, L))
        theta = rand(Uniform(0.0, 2.0*pi))

        # check if the cell is overlapping with others 
        if (noCollision(cells, cparams, cell_ind, a, b, x,
                        y, theta) == 1)
            cells[3*(ind-1)+1:(3*(ind-1)+3)] = Float64[x, y, 0.0]
            cparams[4*(ind-1)+1:(4*(ind-1)+4)] = Float64[a, b, theta, 1]  
            cell_ind = cell_ind + 1  
        end                
    end    

    return cells, cparams, Ovlaps
end    

# Get the overlap indices of all the cells
function find_overlap!(cells, cparams, Ovlaps)

    # iterate through all the cells 
    for i in 1:(size(cells)[1]÷3)
        x1 = cells[3*(i-1)+1]
        y1 = cells[3*(i-1)+2]
        a1 = cparams[4*(i-1)+1]
        b1 = cparams[4*(i-1)+2]
        t1 = cparams[4*(i-1)+3]
        for j in (i+1):(size(cells)[1]÷3)
            x2 = cells[3*(j-1)+1]
            y2 = cells[3*(j-1)+2]
            a2 = cparams[4*(j-1)+1]
            b2 = cparams[4*(j-1)+2]
            t2 = cparams[4*(j-1)+3] 

            # Find the intersection area
            if ellipse_ellipse_overlap(a1, b1, x1, y1, t1, 
                a2, b2, x2, y2, t2) > 0.0
                Ovlaps[i, j] = 1
                Ovlaps[j, i] = 1
            else
                Ovlaps[i, j] = 0
                Ovlaps[j, i] = 0     
            end    
        end
    end    

    return Ovlaps
end    

File: Ellipse_New_Multi_Cell_Model_Numpy/Cell_funcs.py
from ellipses import *
from random_funcs import *

@nb.jit(nopython = True, nogil = True)
def noCollision(cells, cparams, ind, a, b, x, y, theta):

    if ind == 0:
        return True

    x1, y1 = x, y
    a1, b1 = a, b
    t1 = theta
    for i in range(ind):
        x2, y2 = cells[3*i:(3*i+2)]
        a2, b2 = a, b
        t2 = cparams[4*i+2]

        if check_ell_intersection(x1, y1, a1, b1, t1,
                x2, y2, a2, b2, t2):
            return 0

    return 1

@nb.jit(nopython = True, nogil = True)
def preset(a, b, Num):

   # x y dtheta
    cells = np.zeros(Num*3)
    # a b theta phase
    cparams = np.zeros(Num*4)
    Ovlaps = np.zeros((Num, Num)) - 1e8

    for ind in range(Num):

        if ind // (Num//2) == 0 & ind != Num//2:
            theta = pi
        else:
            theta = 0

        cells[3*ind:(3*ind+3)] = np.array([20+2*(a-a/4)*ind, 10, 0])
        cparams[4*ind:(4*ind+4)] = np.array([a, b, theta, 1])

    return cells, cparams, Ovlaps

@nb.jit(nopython = True, nogil = True)
def linear_preset(a, b, Num):

    # x y dtheta
    cells = np.zeros(Num*3)
    # a b theta phase
    cparams = np.zeros(Num*4)
    Ovlaps = np.zeros((Num, Num)) - 1e8

    for ind in range(Num):
        cells[3*ind:(3*ind+3)] = np.array([20+(a-a/4)*ind, 10, 0])
        cparams[4*ind:(4*ind+4)] = np.array([a/2, b, 0, 1])

    return cells, cparams, Ovlaps

@nb.jit(nopython = True, nogil = True)
def two_linear_preset(a, b, Num):

    num = Num//2

    # x y dtheta
    cells = np.zeros(Num*3)
    # a b theta phase
    cparams = np.zeros(Num*4)
    Ovlaps = np.zeros((Num, Num)) - 1e8

    for ind in range(Num):
        cells[3*ind:(3*ind+3)] = np.array([20+2*(a-a/4)*(ind%num), 10+2*(b-b/4)*(ind//num), 0])
        cparams[4*ind:(4*ind+4)] = np.array([a, b, 0, 1])

    return cells, cparams, Ovlaps

@nb.jit(nopython = True, nogil = True)
def custom_preset(a, b, Num):

    Num = int(sqrt(Num))**2
    num = int(sqrt(Num))

    # x y dtheta
    cells = np.zeros(Num*3)
    # a b theta phase
    cparams = np.zeros(Num*4)
    Ovlaps = np.zeros((Num, Num)) - 1e8

    for ind in range(Num):
        cells[3*ind:(3*ind+3)] = np.array([20+2*(a-1)*(ind%num), 10+2*(b-0.5)*(ind//num), 0])
        cparams[4*ind:(4*ind+4)] = np.array([a, b, 0, 1])

    return cells, cparams, Ovlaps

@nb.jit(nopython = True, nogil = True)
def random_cells(L, a, b, Num):

    # x y dtheta
    cells = np.zeros(Num*3)
    # a b theta phase
    cparams = np.zeros(Num*4)
    Ovlaps = np.zeros((Num, Num)) - 1e8

    ind = 0
    while (ind < Num):
        x = uniform_double(0, L, 1)[0]
        y = uniform_double(0, L, 1)[0]
        theta = uniform_double(0, 2*pi, 1)[0]

        # check for collisions
        if noCollision(cells, cparams, ind, a, b, x, y, theta):
            cells[3*ind:(3*ind+3)] = np.array((x, y, 0))
            cparams[4*ind:(4*ind+4)] = np.array((a, b, theta, 1))
            ind += 1

    return cells, cparams, Ovlaps

@nb.jit(nopython = True, nogil = True)
def find_overlaps(cells, cparams, Ovlaps):

    for i in range(cells.shape[0]//3):
        x1, y1 = cells[3*i], cells[3*i+1]
        a1, b1 = cparams[4*i], cparams[4*i+1]
        t1 = cparams[4*i+2]
        for j in range(i+1, cells.shape[0]//3):
            x2, y2 = cells[3*j], cells[3*j+1]
            a2, b2 = cparams[4*j], cparams[4*j+1]
            t2 = cparams[4*j+2]

            if check_ell_intersection(x1, y1, a1, b1, t1,
                                    x2, y2, a2, b2, t2):
                Ovlaps[i][j] = 1
                Ovlaps[j][i] = 1
            else:
                Ovlaps[i][j] = 0
                Ovlaps[j][i] = 0

    return Ovlaps

@nb.jit(nopython = True, nogil = True)
def find_overlaps_ind(cells, cparams, Ovlaps, i):

    x1, y1 = cells[3*i], cells[3*i+1]
    a1, b1 = cparams[4*i], cparams[4*i+1]
    t1 = cparams[4*i+2]
    for j in range(i+1, cells.shape[0]//3):
        x2, y2 = cells[3*j], cells[3*j+1]
        a2, b2 = cparams[4*j], cparams[4*j+1]
        t2 = cparams[4*j+2]

        if check_ell_intersection(x1, y1, a1, b1, t1,
                                x2, y2, a2, b2, t2):
            Ovlaps[i][j] = 1
            Ovlaps[j][i] = 1
        else:
            Ovlaps[i][j] = 0
            Ovlaps[j][i] = 0

    return Ovlaps
